### ISSUE TITLE
Jetty 9.3 SNI host check failing behind Wildfly reverse proxy

### ISSUE THREAD

**Issue by ssmshah1** [eclipse/jetty.project#3282]
I am configuring reverse proxy for Wildfly so that I can access Jetty (Embedded with Solr) through it. SSL is configured on both Wildfly and Jetty and both use same keystore (placed on shared location). However, both are running on separate machines. When I try to access Jetty from browser, through reverse proxy, I get following exception:- HTTP ERROR: 400 Problem accessing /solr. Reason: Host does not match SNI

On further investigation, I found out that during TLS handshake, correct certificate is selected by SNI but while requesting resources from Jetty, host header contains Wildfly's host name (original host header which was sent by browser to reverse proxy) and SNI host check is failed, by Jetty, against selected certificate since it doesn't contain Wildfly's host name in certificate DNS names.

**Comment by ssmshah1** [comment]
I've made following changes to jetty's configuration files:- 

jetty-http.xml
```
<New class="org.eclipse.jetty.server.ServerConnector">
        <Arg name="server"><Ref refid="Server" /></Arg>
        <Arg name="acceptors" type="int"><Property name="solr.jetty.http.acceptors" default="-1"/></Arg>
        <Arg name="selectors" type="int"><Property name="solr.jetty.http.selectors" default="-1"/></Arg>
        <Arg name="factories">
          <Array type="org.eclipse.jetty.server.ConnectionFactory">
            <Item>
              <New class="org.eclipse.jetty.server.HttpConnectionFactory">
                <Arg name="config"><Ref refid="httpConfig" /></Arg>
              </New>
            </Item>
          </Array>
        </Arg>
        <Set name="host"><Property name="jetty.host" /></Set>
        <Set name="port"><Property name="jetty.port" default="8983" /></Set>
        <Set name="idleTimeout"><Property name="solr.jetty.http.idleTimeout" default="50000"/></Set>
        <Set name="soLingerTime"><Property name="solr.jetty.http.soLingerTime" default="-1"/></Set>
        <Set name="acceptorPriorityDelta"><Property name="solr.jetty.http.acceptorPriorityDelta" default="0"/></Set>
        <Set name="selectorPriorityDelta"><Property name="solr.jetty.http.selectorPriorityDelta" default="0"/></Set>
        <Set name="acceptQueueSize"><Property name="solr.jetty.http.acceptQueueSize" default="0"/></Set>
      </New>
```

jetty-https.xml
```
      <New class="org.eclipse.jetty.server.ServerConnector">
        <Arg name="server"><Ref refid="Server" /></Arg>
        <Arg name="acceptors" type="int"><Property name="solr.jetty.ssl.acceptors" default="-1"/></Arg>
        <Arg name="selectors" type="int"><Property name="solr.jetty.ssl.selectors" default="-1"/></Arg>
        <Arg name="factories">
          <Array type="org.eclipse.jetty.server.ConnectionFactory">
            <Item>
              <New class="org.eclipse.jetty.server.SslConnectionFactory">
                <Arg name="next">http/1.1</Arg>
                <Arg name="sslContextFactory"><Ref refid="sslContextFactory"/></Arg>
              </New>
            </Item>
            <Item>
              <New class="org.eclipse.jetty.server.HttpConnectionFactory">
                <Arg name="config"><Ref refid="sslHttpConfig"/></Arg>
              </New>
            </Item>
          </Array>
        </Arg>
        <Set name="host"><Property name="solr.jetty.host" /></Set>
        <Set name="port"><Property name="solr.jetty.https.port" default="8983" /></Set>
        <Set name="idleTimeout"><Property name="solr.jetty.https.timeout" default="50000"/></Set>
        <Set name="soLingerTime"><Property name="solr.jetty.https.soLingerTime" default="-1"/></Set>
        <Set name="acceptorPriorityDelta"><Property name="solr.jetty.ssl.acceptorPriorityDelta" default="0"/></Set>
        <Set name="selectorPriorityDelta"><Property name="solr.jetty.ssl.selectorPriorityDelta" default="0"/></Set>
        <Set name="acceptQueueSize"><Property name="solr.jetty.https.acceptQueueSize" default="0"/></Set>
      </New>
```

jetty-ssl.xml
```
<Configure class="org.eclipse.jetty.util.ssl.SslContextFactory" id="sslContextFactory">
      <Set name="KeyStorePath">
         <Property default="D:\Shared\ssl_setup\cert.keystore" name="solr.jetty.keystore" />
      </Set>
      <Set name="KeyStorePassword">
         <Property default="changeit" name="solr.jetty.keystore.password" />
      </Set>
      <Set name="TrustStorePath">
         <Property default="D:\Shared\ssl_setup\cert.keystore" name="solr.jetty.truststore" />
      </Set>
      <Set name="TrustStorePassword">
         <Property default="changeit" name="solr.jetty.truststore.password" />
      </Set>
      <Set name="NeedClientAuth">
         <Property default="false" name="solr.jetty.ssl.needClientAuth" />
      </Set>
      <Set name="WantClientAuth">
         <Property default="false" name="solr.jetty.ssl.wantClientAuth" />
      </Set>
      <Set name="excludeProtocols">
         <Array type="java.lang.String">
            <Item>SSLv3</Item>
         </Array>
      </Set>
      <Set name="ExcludeCipherSuites">
         <Array type="String">
            <Item>SSL_RSA_WITH_DES_CBC_SHA</Item>
            <Item>SSL_DHE_RSA_WITH_DES_CBC_SHA</Item>
            <Item>SSL_DHE_DSS_WITH_DES_CBC_SHA</Item>
            <Item>SSL_RSA_EXPORT_WITH_RC4_40_MD5</Item>
            <Item>SSL_RSA_EXPORT_WITH_DES40_CBC_SHA</Item>
            <Item>SSL_DHE_RSA_EXPORT_WITH_DES40_CBC_SHA</Item>
            <Item>SSL_DHE_DSS_EXPORT_WITH_DES40_CBC_SHA</Item>
         </Array>
      </Set>
      <New class="org.eclipse.jetty.server.HttpConfiguration" id="sslHttpConfig">
         <Arg>
            <Ref refid="httpConfig" />
         </Arg>
         <Call name="addCustomizer">
            <Arg>
               <New class="org.eclipse.jetty.server.SecureRequestCustomizer" />
            </Arg>
         </Call>
      </New>
      <Set name="certAlias">dssearch_dev</Set>
   </Configure>
```

**Comment by sbordet** [comment]
@ssmshah1 we need more information about domain names that you use from the browser, what's in the certificates and a ASCII diagram of who calls who.

If the browser sends the wildfly host name it means that you put the wildfly host name in the address bar, so it seems normal that the SNI check fails, if you have a certificate that only has the jetty host name.

**Comment by ssmshah1** [comment]
![workflow](https://user-images.githubusercontent.com/46921156/52350525-9a871000-2a4a-11e9-98d9-38ed58705863.png)
During TLS handshake "server_name" is used to select the certificate by Jetty. After successful handshake, when proxy requests for server's resources, the server validates each request by using "HOST" request header (lmkr-4263) which mismatches certificate that it selected during handshake using server name (lmkr-4241). 

Jetty's Certificate:-
![certificate](https://user-images.githubusercontent.com/46921156/52350478-817e5f00-2a4a-11e9-9af5-e2a356bba5aa.PNG)

**Comment by sbordet** [comment]
You have a non-tunneling reverse proxy, so you need to remap the URIs coming from the client to URIs that go to Jetty.
In your reverse proxy you need to replace the host header that the client sent with the host header of the Jetty server.

I'm not sure what you mean by:

> the server validates each request by using "HOST" request header (lmkr-4263) which mismatches certificate that it selected during handshake

Once a TLS connection has been established between the proxy and the server, Jetty does not do any "validation" based on the `Host` header, unless you have virtual hosts configured.
If the proxy needs to open a new TLS connection to the server, it needs to send the right server name.

Seems to me that your problem has nothing to do with Jetty: just do the correct URI remapping in your reverse proxy and everything should work.

**Comment by ssmshah1** [comment]
Agreed. In my scenario I've a limitation that I cannot rewrite "HOST" header so I'll have to go for another workaround and by validation I meant SNI host check that is performed by "SecureRequestCustomizer".

**Comment by sbordet** [comment]
What are you using on the Proxy?

For example, if you use Jetty's `AsyncProxyServlet` you will be able to change the `Host` header.
Jetty's `AsyncProxyServlet` should work in WildFly too - it's entirely based on Servlet API only.

Alternatively, have you tried to write your own `SecureRequestCustomizer`?

**Comment by ssmshah1** [comment]
I've configured KeyCloak for authenticating Jetty. `Host` header is also utilized by KeyCloak and rewriting the header (through reverse proxy) breaks KeyCloak URI redirection. For now, I've added the Wildfly host machine's FQDN to Jetty's certificate to bypass SNI host check failure.
Creating custom `SecureRequestCustomizer ` would require an impact analysis so I'll skip it for now.
Thanks for your help.
